### ISSUE TITLE
Failed while running "make" on Debian Testing

### ISSUE THREAD

**Issue by kruug** [anbox/anbox#147]
```
[ 94%] Linking CXX executable at_parser_tests
/usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `std::_Function_handler<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), AtParser_BasicCommands_Test::TestBody()::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const::{lambda(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)#1}>::_M_invoke(std::_Any_data const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)':
at_parser_tests.cpp:(.text+0x71): undefined reference to `testing::internal::CmpHelperSTRCASEEQ(char const*, char const*, char const*, char const*)'
at_parser_tests.cpp:(.text+0x89): undefined reference to `testing::Message::Message()'
at_parser_tests.cpp:(.text+0xb9): undefined reference to `testing::internal::AssertHelper::AssertHelper(testing::TestPartResult::Type, char const*, int, char const*)'
at_parser_tests.cpp:(.text+0xc4): undefined reference to `testing::internal::AssertHelper::operator=(testing::Message const&) const'
at_parser_tests.cpp:(.text+0xcc): undefined reference to `testing::internal::AssertHelper::~AssertHelper()'
at_parser_tests.cpp:(.text+0xde): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0x104): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0x14e): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0x18d): undefined reference to `testing::internal::AssertHelper::~AssertHelper()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `AtParser_BasicCommands_Test::TestBody()':
at_parser_tests.cpp:(.text+0xd02): undefined reference to `testing::AssertionSuccess()'
at_parser_tests.cpp:(.text+0xd1d): undefined reference to `testing::Message::Message()'
at_parser_tests.cpp:(.text+0xd52): undefined reference to `testing::internal::AssertHelper::AssertHelper(testing::TestPartResult::Type, char const*, int, char const*)'
at_parser_tests.cpp:(.text+0xd5d): undefined reference to `testing::internal::AssertHelper::operator=(testing::Message const&) const'
at_parser_tests.cpp:(.text+0xd65): undefined reference to `testing::internal::AssertHelper::~AssertHelper()'
at_parser_tests.cpp:(.text+0xd77): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0xda0): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0xe19): undefined reference to `testing::internal::IsTrue(bool)'
at_parser_tests.cpp:(.text+0x1079): undefined reference to `testing::internal::EqFailure(char const*, char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
at_parser_tests.cpp:(.text+0x171c): undefined reference to `testing::internal::AssertHelper::~AssertHelper()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `AtParser_BasicCommands_Test::~AtParser_BasicCommands_Test()':
at_parser_tests.cpp:(.text._ZN27AtParser_BasicCommands_TestD2Ev[_ZN27AtParser_BasicCommands_TestD5Ev]+0xf): undefined reference to `testing::Test::~Test()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `AtParser_BasicCommands_Test::~AtParser_BasicCommands_Test()':
at_parser_tests.cpp:(.text._ZN27AtParser_BasicCommands_TestD0Ev[_ZN27AtParser_BasicCommands_TestD5Ev]+0x13): undefined reference to `testing::Test::~Test()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `testing::internal::TestFactoryImpl<AtParser_BasicCommands_Test>::CreateTest()':
at_parser_tests.cpp:(.text._ZN7testing8internal15TestFactoryImplI27AtParser_BasicCommands_TestE10CreateTestEv[_ZN7testing8internal15TestFactoryImplI27AtParser_BasicCommands_TestE10CreateTestEv]+0x17): undefined reference to `testing::Test::Test()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `_GLOBAL__sub_I_at_parser_tests.cpp':
at_parser_tests.cpp:(.text.startup+0xf6): undefined reference to `testing::internal::GetTestTypeId()'
at_parser_tests.cpp:(.text.startup+0x16f): undefined reference to `testing::internal::MakeAndRegisterTestInfo(char const*, char const*, char const*, char const*, testing::internal::CodeLocation, void const*, void (*)(), void (*)(), testing::internal::TestFactoryBase*)'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `testing::internal::scoped_ptr<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >::reset(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
at_parser_tests.cpp:(.text._ZN7testing8internal10scoped_ptrINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE5resetEPS7_[_ZN7testing8internal10scoped_ptrINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEE5resetEPS7_]+0x15): undefined reference to `testing::internal::IsTrue(bool)'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o: In function `testing::internal::scoped_ptr<std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> > >::reset(std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >*)':
at_parser_tests.cpp:(.text._ZN7testing8internal10scoped_ptrINSt7__cxx1118basic_stringstreamIcSt11char_traitsIcESaIcEEEE5resetEPS7_[_ZN7testing8internal10scoped_ptrINSt7__cxx1118basic_stringstreamIcSt11char_traitsIcESaIcEEEE5resetEPS7_]+0x17): undefined reference to `testing::internal::IsTrue(bool)'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o:(.data.rel.ro._ZTI27AtParser_BasicCommands_Test[_ZTI27AtParser_BasicCommands_Test]+0x10): undefined reference to `typeinfo for testing::Test'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o:(.data.rel.ro._ZTV27AtParser_BasicCommands_Test[_ZTV27AtParser_BasicCommands_Test]+0x20): undefined reference to `testing::Test::SetUp()'
CMakeFiles/at_parser_tests.dir/at_parser_tests.cpp.o:(.data.rel.ro._ZTV27AtParser_BasicCommands_Test[_ZTV27AtParser_BasicCommands_Test]+0x28): undefined reference to `testing::Test::TearDown()'
collect2: error: ld returned 1 exit status
tests/anbox/support/CMakeFiles/at_parser_tests.dir/build.make:122: recipe for target 'tests/anbox/support/at_parser_tests' failed
make[2]: *** [tests/anbox/support/at_parser_tests] Error 1
CMakeFiles/Makefile2:2143: recipe for target 'tests/anbox/support/CMakeFiles/at_parser_tests.dir/all' failed
make[1]: *** [tests/anbox/support/CMakeFiles/at_parser_tests.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

Full log: https://paste.debian.net/928393

**Comment by subjectdenied** [comment]
there is a debian fork for anbox here: https://github.com/zhsj/anbox
dev seems to have prebuilt packages too: https://ftp.ustclug.org/~zsj/anbox/

**Comment by kruug** [comment]
Same build issues on that repo.  Looks like it's literally a copy with the Debian changes not pushed.

**Comment by morphis** [comment]
@subjectdenied Can you move the debian bits into a packaging/debian directory and propose a PR? Would like to get them directly included in here.

**Comment by subjectdenied** [comment]
@morphis: fork is not done by me, but i created a pull-request anyways

**Comment by zhsj** [comment]
If you want to build tests, you should setup a proper google test library.
I disabled tests in https://github.com/zhsj/anbox/

@morphis 
I didn't send the pull request since:
* I haven't figured out the correct way to intergrate the googletest.
* dbus-cpp, properties-cpp, process-cpp are not available in Debian(and these libraries are aslo hard to build against the googletest)

**Comment by morphis** [comment]
@zhsj Looks like debian has google-test available as a package: https://packages.debian.org/source/jessie/gtest

The *-cpp libraries are available as deb packages in Ubuntu. Maybe a good source for those.

**Comment by zhsj** [comment]
@morphis 
Yes, these \*-cpp libraries I forked from ubuntu directly and release in github.com/zhsj/\*-cpp

I know googletest is in Debian.
But there are several problems I met.

1. googletest in Debian doesn't ship pre-build binary
    the maintainer of Debian says "Use of precompiled libgtest Not Recommended"
2. find_package(GMock) not work since there's no such cmake rule available in Debian's cmake-data
3. The cmake rule should change to use ExternalProject or other things to build the googletest first. However I never use googletest before and I have no idea after reading the googletest documents.

**Comment by zhsj** [comment]
I get some inspiration form #44 and figure out how to build against googletest when packaging, https://github.com/zhsj/anbox/commit/3026b188bd7cbd00dd962cc8c7333939226a46f6

@morphis How would you like to include these packaging scripts here, some stuff are Debian related.

**Comment by kruug** [comment]
@zhsj If you replace the *-cpp with *-c++, the packages become available on Debian.  It's because of Ubuntu's wonderful standard naming scheme that it doesn't work out of the box.

**Comment by zhsj** [comment]
@kruug not exactly, [dbus-cpp](https://launchpad.net/dbus-cpp) is not the same as [dbus-c++](http://dbus-cplusplus.sourceforge.net/)

**Comment by chuzarski** [comment]
@kruug Under Debian testing I was able to get Anbox to compile and sort of start a container. Anbox says that the container is running. Just like @zhsj I had to disable tests to get it to compile. I also manually compiled and installed *-cpp and friends. 

The only problem I run into now is apparently Android is failing to boot. I cannot figure that out.
`cody@casper:~$ ANBOX_LOG_LEVEL=debug anbox launch --package=org.anbox.appmgr --component=org.anbox.appmgr.AppViewActivity
[DD 2017-04-20 23:30:51] [launch.cpp:96@operator()] Android hasn't fully booted yet. Waiting a bit..
[WW 2017-04-20 23:31:21] [launch.cpp:115@operator()] Stop waiting as we're already waiting for too long. Something is wrong
[WW 2017-04-20 23:31:21] [launch.cpp:116@operator()] with your setup and the container may have failed to boot.
`

@zhsj any thoughts on how to get around this?
